In [154]:
# Importing essential libraries

import pandas as pd # Library for data analysis
import numpy as np # Library for mathematical operations
import seaborn as sns # Library for easy data visualization
import matplotlib.pyplot as plt # Library for custom data visualization
from scipy import stats # Library for performing statistical calculations and analysis

In [155]:
# Declaring path to the file
abosolute_path_to_original_dataset = r"cardekho_dataset_source_kaggle.csv"
# Reading the dataframe using pandas
vehicle_data = pd.read_csv(abosolute_path_data_file)

# Displaying the first 5 observations of the dataset
vehicle_data.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

                     torque  seats  
0            190Nm@ 2000rpm    5.0  
1       250Nm@ 1500-2500rpm    5.0  
2     12.7@ 2,700(kgm@ rpm)    5.0  
3  22.4 kgm at 1750-2750rpm    5.0  
4     11.5@ 4,500(kgm@ rpm)    5.0

In [156]:
# Observing the basics of the dataset
print("Shape of the dataset: ", vehicle_data.shape)
print()
vehicle_data.info() # Displays the information of the dataset: data types and null values

Shape of the dataset:  (8128, 13)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB


**Observation**:
- We are dealing with a very large datset, for the purpose of this project, the dataset will be optimized and made simpler.
- There could be enough dupicate values that has to be removed
- There are 8128 rows and 13 columns in this dataset, this can be reduced to <=1000 observations using random sampling approch 
- Brand name can be extracted from the name column 
- There are no null values in the dataset, however incorrect values maybe present
- For simplicity we will drop some columns: mileage(fuel efficiency per litre), owner (nth owner), max_power, torque, seats

In [157]:
# Removing the dupicate observations from the dataset
print("No. of rows in the dataset before removing the duplicate observations: ", vehicle_data.shape[0])
vehicle_data.drop_duplicates(subset=None, keep='first', inplace=True)

# Displaying the number of rows/observations after removing duplicates
print("No. of rows in the dataset after removing the duplicate observations: ", vehicle_data.shape[0])

No. of rows in the dataset before removing the duplicate observations:  8128
No. of rows in the dataset after removing the duplicate observations:  6926


In [158]:
# Random sampling: Takes the 0.1108% random samples of the entire dataset 
vehicle_data = vehicle_data.sample(frac = 0.13008)

# Number of rows and columns
print("Current shape of the dataset: ", vehicle_data.shape)
# Number of rows of the dataset
print("Total observations after dropping 0.13008% of the orginal observations: " ,vehicle_data.shape[0])

Current shape of the dataset:  (901, 13)
Total observations after dropping 0.13008% of the orginal observations:  901


In [159]:
# Printing each column name in an organized fashion 
print("List of columns in the datset: ")
for each_column in vehicle_data.columns:
    print(each_column)
# Empty line 
print() 
# Total no. of columns in the dataset
print("Current total number of columns in dataset: ", len(vehicle_data.columns))
print()

# Declaring the list of columns to be dropped
columns_to_be_dropped = ["seller_type", "transmission", "owner", "mileage", "max_power", "torque", "seats", "fuel"]
# Prints each column to be dropped in a seperate line
print("List of columns to be dropped now: ")
for each_column in columns_to_be_dropped:
    print(each_column)
print()
# Dropping the listed columns from the dataset
vehicle_data.drop(columns=columns_to_be_dropped, inplace=True)
# Prints the current shape of the dataframe after dropping
print("Shape of the dataframe after dropping 8 columns: ", vehicle_data.shape)
# Prints the current no. of coumns after dropping
print("Total number of columns after dropping 8 columns: ", vehicle_data.shape[1])

List of columns in the datset: 
name
year
selling_price
km_driven
fuel
seller_type
transmission
owner
mileage
engine
max_power
torque
seats

Current total number of columns in dataset:  13

List of columns to be dropped now: 
seller_type
transmission
owner
mileage
max_power
torque
seats
fuel

Shape of the dataframe after dropping 8 columns:  (901, 5)
Total number of columns after dropping 8 columns:  5


In [160]:
# Displays the basics of our modified dataframe
vehicle_data.info()

# Displays the current dataframe 
vehicle_data.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 901 entries, 6709 to 5391
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   name           901 non-null    object
 1   year           901 non-null    int64 
 2   selling_price  901 non-null    int64 
 3   km_driven      901 non-null    int64 
 4   engine         877 non-null    object
dtypes: int64(3), object(2)
memory usage: 42.2+ KB


name  year  selling_price  km_driven   engine
6709           Tata Xenon XT EX 4X2  2014         440000     110000  2179 CC
7213  Chevrolet Optra Magnum 2.0 LS  2007         160000     130000  1991 CC
6336       Maruti Wagon R VXI BS IV  2012         250000      65000   998 CC
7118              Tata Sumo Gold GX  2013         350000     120000  2956 CC
2187                   Honda City E  2012         560000      40000  1497 CC

In [161]:
# Extracting the brand name from the name column
vehicle_data["Vehicle_brand"] = vehicle_data["name"].apply(lambda x: x.split(" ")[0])
# Dropping the name column from the dataframe
vehicle_data.drop("name", inplace=True,axis = 1)

# Displaying the current dataframe
vehicle_data.head()

year  selling_price  km_driven   engine Vehicle_brand
6709  2014         440000     110000  2179 CC          Tata
7213  2007         160000     130000  1991 CC     Chevrolet
6336  2012         250000      65000   998 CC        Maruti
7118  2013         350000     120000  2956 CC          Tata
2187  2012         560000      40000  1497 CC         Honda

In [162]:
# Extracting the CC number from the engine column 
vehicle_data["Cubic_capacity"] = vehicle_data['engine'].astype('str').str.split(expand=True)[0]
# Converting it into integer type data
vehicle_data['Cubic_capacity'] = vehicle_data['Cubic_capacity'].astype('float64')

# Dropping the engine column from the dataframe
vehicle_data.drop("engine", inplace=True, axis = 1)

# Displaying the current dataframe
vehicle_data.head()

year  selling_price  km_driven Vehicle_brand  Cubic_capacity
6709  2014         440000     110000          Tata          2179.0
7213  2007         160000     130000     Chevrolet          1991.0
6336  2012         250000      65000        Maruti           998.0
7118  2013         350000     120000          Tata          2956.0
2187  2012         560000      40000         Honda          1497.0

In [172]:
# Current information about the dataset  
vehicle_data.info()

# Displaying the number of missing values in the Cubic_capicity column
print() # empty line
print("No. of missing values(NA) in Cubic_capicity column: ", vehicle_data["Cubic_capacity"].isna().sum())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 901 entries, 6709 to 5391
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   year            901 non-null    int64  
 1   selling_price   901 non-null    int64  
 2   km_driven       901 non-null    int64  
 3   Vehicle_brand   901 non-null    object 
 4   Cubic_capacity  877 non-null    float64
dtypes: float64(1), int64(3), object(1)
memory usage: 42.2+ KB

No. of missing values(NA) in Cubic_capicity column:  24


In [167]:
# Handling the missing values in the dataset

vehicle_data["Cubic_capacity"].isna().sum()


24

In [164]:
# Changing the position and the names of the columns of the dataset


# Saving it as a final dataset in directory

